src/coprocessor/dag/batch/executors/stream_aggr_executor.rs

```diff
@@ -0,0 +1,572 @@
+        ensure_columns_decoded(
+            &context.cfg.tz,
+            &entities.each_aggr_exprs,
+            src_schema,
+            &mut input,
+        )?;
+        let group_by_results = eval_exprs(context, &self.group_by_exps, src_schema, &input)?;
+        let aggr_expr_results = eval_exprs(context, &entities.each_aggr_exprs, src_schema, &input)?;
+
+        // Stores input references, clone them when needed
+        let mut group_key_ref = Vec::with_capacity(group_by_len);
+        let mut group_start_row = 0;
+        for row_index in 0..rows_len {
+            for group_by_result in &group_by_results {
+                // Unwrap is fine because we have verified the group by expression before.
+                let group_column = group_by_result.vector_value().unwrap();
+                group_key_ref.push(group_column.get_scalar_ref(row_index));
+            }
+            match self.keys.rchunks_exact(group_by_len).next() {
+                Some(current_key) if &group_key_ref[..] == current_key => {
+                    group_key_ref.clear();
+                }
+                _ => {
+                    // Update the complete group
+                    if row_index > 0 {
+                        update_current_states(
+                            context,
+                            &mut self.states,
+                            aggr_fn_len,
+                            &aggr_expr_results,
+                            group_start_row,
+                            row_index,
+                        )?;
+                    }
+
+                    // create a new group
+                    group_start_row = row_index;
+                    self.keys
+                        .extend(group_key_ref.drain(..).map(ScalarValueRef::to_owned));
+                    for aggr_fn in &entities.each_aggr_fn {
+                        self.states.push(aggr_fn.create_state());
+                    }
+                }
+            }
+        }
+        // Update the current group with the remaining data
+        update_current_states(
+            context,
+            &mut self.states,
+            aggr_fn_len,
+            &aggr_expr_results,
+            group_start_row,
+            rows_len,
+        )?;
+
+        Ok(())
+    }
+
+    /// Note that the partial group is in count.
+    #[inline]
+    fn groups_len(&self) -> usize {
+        self.keys
+            .len()
+            .checked_div(self.group_by_exps.len())
+            .unwrap_or(0)
+    }
+
+    #[inline]
+    fn iterate_each_group_for_partial_aggregation(
+        &mut self,
+        entities: &mut Entities<Src>,
+        src_is_drained: bool,
+        mut iteratee: impl FnMut(&mut Entities<Src>, &[Box<dyn AggrFunctionState>]) -> Result<()>,
+    ) -> Result<Vec<LazyBatchColumn>> {
+        let number_of_groups = if src_is_drained {
+            AggregationExecutorImpl::<Src>::groups_len(self)
```

Comment:
why not `self.groups_len()`?

Reply:
The compiler complains type `Src` cannot be determined...

Reply:
Looks like some wired bugs..

Reply:
@kennytm  Needs help 🙃

Reply:
@breeswish @sticnarf  Not totally a bug, `BatchStreamAggregationImpl` implements `AggregationExecutorImpl<Src>` for all `Src`, so the compiler doesn't know which `Src` to use when you call `group_len()`. Minimal reproduction:

```rust
trait A<S> {
    fn g(&self) -> usize;
    fn i(&mut self);
}

struct B;

impl<S> A<S> for B {
    fn g(&self) -> usize { 
        std::mem::size_of::<S>() 
    }
    fn i(&mut self) {
        println!("{}", self.g());
    }
}
```